File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int inDegree[1000001];
int outDegree[1000001];
bool visited[1000001];

vector<int> adj[1000001];

vector<int> solution(vector<vector<int>> edges) {
    vector<int> answer;
    
    // 제일 큰 번호
    int N = 0;
    
    // 간선 체크
    for(int i = 0; i < edges.size(); ++i){
        inDegree[edges[i][1]]++;
        outDegree[edges[i][0]]++;
        
        // 가장 큰 번호 체크
        if(edges[i][0] > N) N = edges[i][0];
        if(edges[i][1] > N) N = edges[i][1];
        
        // 간선 리스트
        adj[edges[i][0]].push_back(edges[i][1]);
    }
    
    int newPoint = 0;
    for(int i = 1; i <= N; ++i){
        if(outDegree[i] >= 2 && inDegree[i] == 0){
            newPoint = i;
            break;
        }
    }
    
    int donut = 0, stick = 0, eight = 0;
    
    // 새로 생긴 점에서 출발해서 탐색
    for(int i = 0; i < adj[newPoint].size(); ++i){
        // 점의 개수와 간선의 개수를 체크하자
        int start = adj[newPoint][i];
        
        queue<int> q;
        
        q.push(start);
        visited[start] = true;
        
        int pCnt = 1; int lCnt = 0;
        
        while(!q.empty()){
            int cur = q.front();
            q.pop();
            
            for(int i = 0; i < adj[cur].size(); ++i){
                lCnt++;
                
                int next = adj[cur][i];
                if(!visited[next]){
                    visited[next] = true;
                    pCnt++;
                    q.push(next);
                }
            }
        }
        
        if(pCnt == lCnt) donut++;
        else if(pCnt == lCnt + 1) stick++;
        else if(pCnt == lCnt - 1) eight++;
    }
    
    answer.push_back(newPoint);
    answer.push_back(donut);
    answer.push_back(stick);
    answer.push_back(eight);
    
    return answer;
}

    // 단방향 그래프
    
    // 도넛 - 사이클
    
    // 막대 - 사이클인데 간선 1개 없음
    
    // n + 1 사이클 두개 / 1점 공유
    
    // 새로 생긴 친구
    // 자기한테 들어오는 게 1개도 없고
    // 나가는게 2개 이상
    
    // 빼고 탐색하기

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.*;

class Solution {
    static int[] inDegree = new int[1000001];
    static int[] outDegree = new int[1000001];
    static boolean[] visited = new boolean[1000001];
    static boolean[] exists = new boolean[1000001];
    static List[] adjList = new ArrayList[1000001];

    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        // 도넛 -> 순회 -> 1 out 1 in
        // 막대 -> 1자 -> 1 out / 1 in / 1 out 1 in
        // 8자 -> 1 out 1 in / 2 out 2 in
        
        // 들어오는 점이 없다 + 나가는 점이 1개 초과 -> 추가된 정점?
        
        for(int i = 0; i < 1000001; ++i){
            adjList[i] = new ArrayList<Integer>();
        }
        
        for(int i = 0; i < edges.length; ++i){
            outDegree[edges[i][0]]++;
            inDegree[edges[i][1]]++;
            
            exists[edges[i][0]] = exists[edges[i][1]] = true;
            
            adjList[edges[i][0]].add(edges[i][1]);
            adjList[edges[i][1]].add(edges[i][0]);
        }
        
        for(int i = 1; i <= 1000000; ++i){
            if(inDegree[i] == 0 && outDegree[i] > 1) {
                answer[0] = i;
                break;
            }
        }
        
        for(int i = 1; i <= 1000000; ++i) {
            if(visited[i] || !exists[i] ||i == answer[0]) continue;
            visited[i] = true;
            
            int node_count = 0;
            int edge_count = 0;
            
            Queue<Integer> q = new ArrayDeque<>();
            q.offer(i);
            
            while(!q.isEmpty()) {
                int cur = q.poll();
                
                node_count++;
                edge_count += inDegree[cur] + outDegree[cur];
                
                for(int j = 0; j < adjList[cur].size(); ++j){
                    int nx = (int)adjList[cur].get(j);
                    if(visited[nx] || nx == answer[0]) continue;
                    visited[nx] = true;
                    q.offer(nx);
                }
            }
            
            if (node_count == (edge_count - 1) / 2) answer[1]++;
            else if (node_count - 1 == (edge_count - 1) / 2) answer[2]++;
            else answer[3]++;
        }
        
        // answer :  생성한 정점, 도넛, 막대, 8자
        
        return answer;
    }
}